### PR TITLE
added tests for, and fixes for, discrete mask bugs

### PR DESCRIPTION
**Pull request by josephdviviano**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description
Fixes a bug where masks did not always end up on the right device, adds tests.

### CODE REVIEW

File: src/gfn/states.py [hunk 1]
```diff
@@ -602,8 +603,12 @@ def stack(cls, states: Sequence[DiscreteStates]) -> DiscreteStates:
...
         out = super().stack(states)
         assert isinstance(out, DiscreteStates)
-        out.forward_masks = torch.stack([s.forward_masks for s in states], dim=0)
-        out.backward_masks = torch.stack([s.backward_masks for s in states], dim=0)
+        out.forward_masks = torch.stack([s.forward_masks for s in states], dim=0).to(
+            out.device
+        )
+        out.backward_masks = torch.stack([s.backward_masks for s in states], dim=0).to(
+            out.device
+        )
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/states.py:611]: i did this for safety. I don't think moving a tensor from -> to the same device is a no-op.
